### PR TITLE
GB13 search

### PR DESCRIPTION
**Pull request by kris-kolosek**

### CODE REVIEW

File: db/migrate/20180119145644_devise_invitable_add_to_users.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      t.index      :invitations_count
+      t.index      :invitation_token, unique: true # for invitable
+      t.index      :invited_by_id
+    end
+  end
+
+  def down
+    change_table :users do |t|
+      t.remove_references :invited_by, polymorphic: true
+      t.remove :invitations_count, :invitation_limit, :invitation_sent_at, :invitation_accepted_at, :invitation_token, :invitation_created_at
```
Comment (sniffler) [inline db/migrate/20180119145644_devise_invitable_add_to_users.rb]: Line is too long. [141/80]

File: app/services/text_formatter.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+require "redcarpet"
+
+module TextFormatter
+
```
Comment (sniffler) [inline app/services/text_formatter.rb]: Extra empty line detected at module body beginning.

File: app/models/user.rb [hunk 3]
```diff
@@ -17,6 +18,10 @@ class User < ApplicationRecord
   private
 
   def set_role
-    self.role = Role.new(user: self, role: Role::ADMIN)
+    if User.count.zero?
```
Comment (sniffler) [inline app/models/user.rb]: Use the return of the conditional for variable assignment and comparison.

File: app/controllers/invitations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      format.html do
+        super
+      end
+    end
+  end
+
+  protected
+
+  def update_sanitized_params
+    devise_parameter_sanitizer.permit(:accept_invitation, keys: [:password, :invitation_token])
```
Comment (sniffler) [inline app/controllers/invitations_controller.rb]: Use %i or %I for an array of symbols.<br>Line is too long. [95/80]

File: app/controllers/invitations_controller.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  before_action :update_sanitized_params, only: :update
+
+  # PUT /resource/invitation
+  def update
+    respond_to do |format|
+      format.js do
+        invitation_token = Devise.token_generator.digest(resource_class, :invitation_token, update_resource_params[:invitation_token])
+        self.resource = resource_class.where(invitation_token: invitation_token).first
+        resource.skip_password = true
+        resource.update_attributes update_resource_params.except(:invitation_token)
```
Comment (sniffler) [inline app/controllers/invitations_controller.rb]: Line is too long. [83/80]

File: app/controllers/invitations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
+class InvitationsController < Devise::InvitationsController
+  before_action :update_sanitized_params, only: :update
+
+  # PUT /resource/invitation
+  def update
+    respond_to do |format|
+      format.js do
+        invitation_token = Devise.token_generator.digest(resource_class, :invitation_token, update_resource_params[:invitation_token])
+        self.resource = resource_class.where(invitation_token: invitation_token).first
```
Comment (sniffler) [inline app/controllers/invitations_controller.rb]: Line is too long. [86/80]

File: app/controllers/invitations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+class InvitationsController < Devise::InvitationsController
+  before_action :update_sanitized_params, only: :update
+
+  # PUT /resource/invitation
+  def update
+    respond_to do |format|
+      format.js do
+        invitation_token = Devise.token_generator.digest(resource_class, :invitation_token, update_resource_params[:invitation_token])
```
Comment (sniffler) [inline app/controllers/invitations_controller.rb]: Line is too long. [134/80]

File: app/assets/stylesheets/pages/_story.scss [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+.edit-story-content {
+  background-color: $white;
+  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen,Ubuntu,Cantarell,Droid Sans,Helvetica Neue,sans-serif;
+  left: 300px;
+  max-width: 1200px;
+  padding: 24px 4vw 3vw;
+  position: absolute;
+  width: 100%;
+
+  .form-title {
```
Comment (sniffler) [inline app/assets/stylesheets/pages/_story.scss]: Rule set contains (11/10) properties

File: app/assets/stylesheets/pages/_story.scss [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      font-size: 20px;
+      font-weight: 700;
+      letter-spacing: .8px;
+      line-height: 1.15em;
+      min-height: auto;
+      overflow: hidden;
+      width: 100%;
+    }
+
+    .form-text {
```
Comment (sniffler) [inline app/assets/stylesheets/pages/_story.scss]: Selector should have depth of applicability no greater than 2, but was 3

File: app/assets/stylesheets/pages/_story.scss [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+  }
+
+  .story {
+    margin-bottom: 10px;
+
+    .form-title {
+      color: $black;
+      font-size: 20px;
+      font-weight: 700;
+      letter-spacing: .8px;
```
Comment (sniffler) [inline app/assets/stylesheets/pages/_story.scss]: `.8` should be written with a leading zero as `0.8`

File: app/controllers/stories_controller.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    @story = Story.new
+  end
+
+  # POST /stories
+  def create
+    @story = Story.new(story_params)
+    @story.user = current_user
+
+    if @story.save
+      flash[:success] = "Story is successfully created."
```
Comment (salac-kolosek) [inline app/controllers/stories_controller.rb]: You can define add_flash_types :success, :error in application controller and then use it like redirect_to team_members_path, success: "Profile is removed successfully" for example.

File: app/models/story.rb [hunk 12]
```diff
@@ -0,0 +1,28 @@
+class Story < ApplicationRecord
+  belongs_to :user
+  before_save :markdown
+
+  def markdown
```
Comment (salac-kolosek) [inline app/models/story.rb]: Put markdown in private methods.

File: app/models/user.rb [hunk 13]
```diff
@@ -17,6 +18,10 @@ class User < ApplicationRecord
...
+      self.role = Role.new(user: self, role: Role::ADMIN)
```
Comment (salac-kolosek) [inline app/models/user.rb]: Make role like this build_role(status: Role::Admin)

File: app/views/devise/mailer/invitation_instructions.html.erb [hunk 14]
```diff
@@ -0,0 +1,11 @@
+<p><%= t("devise.mailer.invitation_instructions.hello", email: @resource.email) %></p>
+
+<p><%= t("devise.mailer.invitation_instructions.someone_invited_you", url: root_url) %></p>
+
+<p><%= link_to t("devise.mailer.invitation_instructions.accept"), accept_invitation_url(@resource, :invitation_token => @token) %></p>
```
Comment (salac-kolosek) [inline app/views/devise/mailer/invitation_instructions.html.erb]: Use rails 5 syntax (invitation_token: value) for example. Be consistent throughout code.

File: spec/factories/profiles.rb [hunk 15]
```diff
@@ -1,5 +1,6 @@
 FactoryGirl.define do
   factory :profile do
+    user
     full_name "Test User"
```
Comment (salac-kolosek) [inline spec/factories/profiles.rb]: You can use sequence(:full_name) to every time create a profile with a different full_name.

File: app/services/text_formatter.rb [hunk 16]
```diff
@@ -0,0 +1,8 @@
+require "redcarpet"
+module TextFormatter
+
```
Comment (sniffler) [inline app/services/text_formatter.rb]: Extra empty line detected at module body beginning.

File: app/models/story.rb [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  def self.search(title)
+    if title.present?
+      where("title ilike ?", "#{title}%")
+    else
+      all
+    end
+  end
+
```
Comment (sniffler) [inline app/models/story.rb]: Trailing whitespace detected.

File: app/assets/stylesheets/pages/_story.scss [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+    border-top: 1px solid;
+    color: $light-gray;
+    margin-top: 25px;
+    padding-top: 10px;
+  }
+
+  .story {
+    margin-bottom: 10px;
+
+    .form-title {
```
Comment (sniffler) [inline app/assets/stylesheets/pages/_story.scss]: Selector should have depth of applicability no greater than 2, but was 3

File: app/models/story.rb [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+  def markdown
```
Comment (salac-kolosek) [inline app/models/story.rb]: Missing private declaration before markdown.

Comment (kolosek) [reply app/models/story.rb]: @salac-kolosek Great catch!

File: app/models/user.rb (hunk 3 above)
Comment (kolosek) [reply app/models/user.rb]: There is a better way to do this. User.count will count ALL users in system. Even when we have 1 Million!
  def make_editor
    self.roles << Role.editor
  end
Take a look how it's done in Applier - https://github.com/kolosek/applier/blob/development/app/models/user.rb

File: app/views/application/_index.html.erb [hunk 20]
```diff
@@ -2,7 +2,7 @@
   <header class="layout-header">
     <h2 class="title">Your stories</h2>
     <section class="header-actions">
-      <a href="#" class="btn actionbtn">New story</a>
+      <a href="/stories/new" class="btn actionbtn">New story</a>
```
Comment (kolosek) [inline app/views/application/_index.html.erb]: Please update links

File: app/views/devise/mailer/invitation_instructions.html.erb (hunk 14 above)
Comment (kolosek) [reply app/views/devise/mailer/invitation_instructions.html.erb]: Correction, that's Ruby 1.9+ syntax, not Rails, but correct, be consistent throughout code.

File: config/initializers/devise.rb [hunk 21]
```diff
@@ -8,25 +8,25 @@
...
   # by default. You can change it below and use your own secret key.
-  # config.secret_key = '568f796722b553f522ac5b1aea90c6ca8fc88ed72d66fb254046cf6eeacf9fe205535f074192103f97cacee654b774942c278468801cd2c9e85b15d7b9a1370c'
+  # config.secret_key = "568f796722b553f522ac5b1aea90c6ca8fc88ed72d66fb254046cf6eeacf9fe205535f074192103f97cacee654b774942c278468801cd2c9e85b15d7b9a1370c"
 
   # ==> Mailer Configuration
   # Configure the e-mail address which will be shown in Devise::Mailer,
   # note that it will be overwritten if you use your own mailer class
   # with default "from" parameter.
-  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
+  config.mailer_sender = ENV["SENDER_EMAIL"]
```
Comment (kolosek) [inline config/initializers/devise.rb]: Put this in Config file. Ask Sale where. This is good, but there is even better way to do that.

File: spec/factories/users.rb [hunk 22]
```diff
@@ -1,6 +1,6 @@
 FactoryGirl.define do
   factory :user do
-    email "test@example.com"
+    sequence(:email) { |n| "name#{n}@example.com" }
```
Comment (kolosek) [inline spec/factories/users.rb]: yes, much better.

File: app/views/application/_index.html.erb (hunk 20 above)
Comment (kris-kolosek) [reply app/views/application/_index.html.erb]: Updated them, thank you for noticing.

File: app/models/user.rb (hunk 3 above)
Comment (kris-kolosek) [reply app/models/user.rb]: I will fix it, thank you.

**Pull request merged**